drivers/common/virtio/virtio.c

```diff
@@ -0,0 +1,339 @@
+	if (!hw->vqs) {
+		PMD_INIT_LOG(ERR, "Failed to alloc virtio device queues.");
+		return -ENOMEM;
+	}
+
+	for (i = 0; i < nr_vq; i++) {
+		vq = rte_zmalloc_socket(NULL, sizeof(struct virtqueue), RTE_CACHE_LINE_SIZE,
+					VTPCI_DEV(hw)->device.numa_node);
+		if (vq == NULL) {
+			PMD_INIT_LOG(ERR, "Failed to alloc virtio queue.");
+			virtio_dev_free_queues(vpdev,nr_vq);
+			return -ENOMEM;
+		}
+		hw->vqs[i] = vq;
+		vq->hw = hw;
+		vq->vq_queue_index = i;
+	}
+
+	return 0;
+}
+
+void
+virtio_dev_get_features(struct virtio_pci_dev *vpdev,uint64_t *features)
+{
+	struct virtio_hw *hw = NULL;
+
+	hw = &vpdev->hw;
+	*features = hw->guest_features;
+}
+
+uint64_t
+virtio_dev_set_features(struct virtio_pci_dev *vpdev,uint64_t features)
+{
+	struct virtio_hw *hw = NULL;
+
+	hw = &vpdev->hw;
+	return virtio_dev_negotiate_features(hw, features);
+}
+
+int
+virtio_dev_get_notify_area(struct virtio_pci_dev *vpdev, int qid, uint64_t *offset, uint64_t *size)
+{
+	struct virtio_hw *hw = NULL;
+	struct vfio_region_info reg = { .argsz = sizeof(reg) };
+	int ret;
+
+	reg.index = vpdev->notify_bar;
+	ret = ioctl(vpdev->vfio_dev_fd, VFIO_DEVICE_GET_REGION_INFO, &reg);
+	if (ret) {
+		PMD_INIT_LOG(ERR, "Can not get device region info: %s",
+				strerror(errno));
+		return -1;
+	}
+
+	hw = &vpdev->hw;
+	//TO_DO: check whether notify_addr is 4k aligned
+	*offset = (uint8_t*)hw->vqs[qid]->notify_addr - (uint8_t*)VTPCI_DEV(hw)->mem_resource[vpdev->notify_bar].addr + reg.offset;
+	//nofiy area should 4k aligned, so should not just one doorbell size
+	*size = 0x1000;
+	return 0;
+
+}
+
+int
+virtio_dev_set_queue(struct virtio_pci_dev *vpdev, int index, struct virtio_dev_vring_info *vring_info)
+{
+	struct virtio_hw *hw = NULL;
+	struct virtqueue *hw_vq;
+	unsigned int size;
+
+	hw = &vpdev->hw;
+	hw_vq = hw->vqs[index];
+
+	hw_vq->vq_ring_mem  = vring_info->desc;
+	hw_vq->vq_avail_mem = vring_info->avail;
+	hw_vq->vq_used_mem  = vring_info->used;
+	hw_vq->vq_nentries  = vring_info->size;
+
+	size = vring_size(hw, vring_info->size, VIRTIO_VRING_ALIGN);
+	hw_vq->vq_ring_size = RTE_ALIGN_CEIL(size, VIRTIO_VRING_ALIGN);
+
+	if (VIRTIO_OPS(hw)->setup_queue(hw, hw_vq) < 0) {
+			PMD_INIT_LOG(ERR, "Setup_queue failed");
+			return -EINVAL;
+	}
+
+	return 0;
+}
+
+void
+virtio_dev_del_queue(struct virtio_pci_dev *vpdev, int index)
+{
+	struct virtio_hw *hw = NULL;
+	struct virtqueue *hw_vq;
+
+	hw = &vpdev->hw;
+	hw_vq = hw->vqs[index];
+
+	VIRTIO_OPS(hw)->del_queue(hw, hw_vq);
+}
+
+int
+virtio_dev_interrupts_enable(struct virtio_pci_dev *vpdev, int nr_vring)
+{
+	struct virtio_hw *hw = &vpdev->hw;
+	int i;
+
+	VIRTIO_OPS(hw)->intr_detect(hw);
+	if(vpdev->msix_status != VIRTIO_MSIX_ENABLED) {
+		PMD_INIT_LOG(ERR, "Error MSI-X not enabled,status: %d", vpdev->msix_status);
+		return -1;
+	}
+
+	if (VIRTIO_OPS(hw)->set_config_irq(hw, 0) ==
+			VIRTIO_MSI_NO_VECTOR) {
+		PMD_INIT_LOG(ERR, "Failed to set config vector");
+		return -1;
```

Comment:
see if you can error code from this callback and use that error code instead of -1.

if not, do -EINVAL.